src/ui/Distance/DistanceResources.ts

```diff
@@ -0,0 +1,245 @@
+   * 
+   * @returns {Promise<IPosition | void>} Promise for the last resolved position value.
+   */
+  public getLastPositionRequest(): Promise<IPosition | void> {
+    return this.lastPositionRequest || Promise.reject('No position request was executed yet.');
+  }
+
+  private onAfterComponentsInitialization(): void {
+    const args: IResolvingPositionEventArgs = {
+      providers: []
+    };
+
+    this.bind.trigger(this.element, DistanceEvents.onResolvingPosition, args);
+
+    this.lastPositionRequest = this.tryGetPositionFromProviders(args.providers)
+      .then(position => {
+        if (!!position) {
+          this.setPosition(position.lat, position.long);
+        } else {
+          this.triggerDistanceNotSet();
+        }
+        return position;
+      })
+      .catch(error => {
+        this.logger.error('An error occured when trying to resolve the current position.', error);
+        this.triggerDistanceNotSet();
+      });
+  }
+
+  private tryGetPositionFromProviders(providers: IPositionProvider[]): Promise<IPosition | void> {
+    return new Promise<IPosition>((resolve, reject) => {
+      const tryNextProvider = () => {
+        if (providers.length > 0) {
+          const provider = providers.shift();
+          provider
+            .getPosition()
+            .then(position => {
+              if (!!position.lat && !!position.long) {
+                resolve(position);
+              } else {
+                tryNextProvider();
+              }
+            })
+            .catch(error => {
+              this.logger.warn('An error occured when tring to resolve the position within a position provider.', error);
+              tryNextProvider();
+            });
+        } else {
+          resolve();
+        }
+      };
+      tryNextProvider();
+    });
+  }
+
+  private triggerDistanceNotSet(): void {
+    this.logger.warn(
+      "None of the given position providers could resolve the current position. The distance field will not be calculated and the distance components will be disabled until the next call to 'setDistance'."
+    );
+    this.bind.trigger(this.element, DistanceEvents.onPositionNotResolved, {});
+  }
+
+  private registerDistanceQuery(): void {
+    this.bind.onRootElement(QueryEvents.buildingQuery, (args: IBuildingQueryEventArgs) => {
+      const geoQueryFunction: IQueryFunction = {
+        function: this.getConvertedUnitsFunction(
+          `dist(${this.options.latitudeField}, ${this.options.longitudeField}, ${this.latitude}, ${this.longitude})`
+        ),
+        fieldName: this.options.distanceField
+      };
+      if (args && args.queryBuilder) {
+        args.queryBuilder.queryFunctions.push(geoQueryFunction);
+        this.enableDistanceComponents();
+      }
+    });
+  }
+
+  private enableDistanceComponents(): void {
+    const rootDomElement = new Dom(this.root);
+
+    rootDomElement.findAll(`.${this.options.disabledDistanceCssClass}`).forEach(element => {
```

Comment:
What are the components that needs to be disabled/enabled when the distance component is disabled/enabled ?

I don't think it should be the distance component that finds what needs to be activated using a css class.

Components that needs to disable/enable  should listen to the appropriate event, and act on their own.

That way you dont have to rely on a flimsy link (the css class), and try/catching the enable/disable mechanism.

Reply:
We used that for Facets and Sort that rely on this, since they are generic Facets and Sort.

I think it is easier to configure a class